colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -472,19 +496,25 @@ def backward_b_step(
         # calculate bwd b step ; only dx = w*dy;
 
         # Retain the grad on the input_obj.
-        tree_map(retain_grad, input_obj)
+        if input_obj is None:
+            return None
+        else:
+            tree_map(retain_grad, input_obj)
+        input_obj_ = input_obj["hidden_states"]
 
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # loss backward; output_obj is loss; so output_obj_grad should be None
             assert output_obj_grad is None
-
+            output_obj_ = output_obj
+        else:
+            output_obj_ = output_obj["hidden_states"]
```

Comment:
remove hardcode

Reply:
> remove hardcode

Check tensor.require grad in input_obj; if True, append output_obj[key] & output_grad_obj[key] to list, else pass;

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/349272c71fa9d30c404ca29b394d7e79bfcd2fd0;